File: motornoencoder.py
from machine import Pin, PWM
from time import sleep_ms, ticks_ms, ticks_diff
import micropython
micropython.alloc_emergency_exception_buf(100)

LFWD = PWM(Pin(2))
RFWD = PWM(Pin(4))
RREV = PWM(Pin(5))
LREV = PWM(Pin(3))
freq = 1000
LFWD.freq(freq)
RFWD.freq(freq)
LREV.freq(freq)
RREV.freq(freq)

RSideInt = Pin(14, Pin.IN, Pin.PULL_UP)
LSideInt = Pin(15, Pin.IN, Pin.PULL_UP)

last_ldir = None
last_rdir = None
reverse_delay_ms = 300

overcurrentFlagRight = False
overcurrentFlagLeft = False
oc_count_r = 0
oc_count_l = 0
OC_THRESHOLD = 150  # must see fault for 150 consecutive 1ms ticks

def check_overcurrent():
    global overcurrentFlagRight, overcurrentFlagLeft
    global oc_count_r, oc_count_l
# 
#     if RSideInt.value() == 1:  # fault = pin pulled low
#         oc_count_r += 1
#         if oc_count_r >= OC_THRESHOLD:
#             overcurrentFlagRight = True
#     else:
#         oc_count_r = 0
#         overcurrentFlagRight = False

    if LSideInt.value() == 1:
        oc_count_l += 1
        if oc_count_l >= OC_THRESHOLD:
            overcurrentFlagLeft = True
    else:
        oc_count_l = 0
        overcurrentFlagLeft = False

def stop_all():
    LFWD.duty_u16(0)
    RFWD.duty_u16(0)
    LREV.duty_u16(0)
    RREV.duty_u16(0)

def move(speed, ldirection, rdirection):
    global last_ldir, last_rdir

    if (last_ldir is not None and ldirection != last_ldir) or \
       (last_rdir is not None and rdirection != last_rdir):
        stop_all()
        sleep_ms(reverse_delay_ms)

    if speed == 0:
        stop_all()
        return

    duty = int(speed * 65535)

    if overcurrentFlagRight:
        LFWD.duty_u16(0)
        LREV.duty_u16(0)
        print("Right Side Overcurrent")
    else:
        if ldirection:
            LFWD.duty_u16(duty)
        else:
            LREV.duty_u16(duty)

    if overcurrentFlagLeft:
        RFWD.duty_u16(0)
        RREV.duty_u16(0)
        print("Left Side Overcurrent")
    else:
        if rdirection:
            RFWD.duty_u16(duty)
        else:
            RREV.duty_u16(duty)

    last_ldir = ldirection
    last_rdir = rdirection

i = 0
did_forward = False

try:
    last_action = 0
    while True:
        now = ticks_ms()
        check_overcurrent()
        sleep_ms(1)

        if overcurrentFlagRight or overcurrentFlagLeft:
            stop_all()
            print("Overcurrent - stopped")
            continue

        if ticks_diff(now, last_action) > 500:
            last_action = now
            i += 1
            if i > 40:
                pass
            elif i > 20:
                if not did_forward:
                    move(0.5, True, True)
                    print("forward")
                    did_forward = True
            else:
                move(0, False, False)
                print("Time till Start:", i)

except KeyboardInterrupt:
    stop_all()

File: motornoencoder.c
#include "motornoencoder.h"
#include "global.h"
#include <stdio.h>
#include <math.h>

#include "hardware/gpio.h"
#include "hardware/pwm.h"
#include "pico/stdlib.h"

static bool motor_noencoder_fault_pin_valid(unsigned int pin) {
    return pin != MOTOR_NOENCODER_NO_FAULT_PIN;
}

static void motor_noencoder_setup_pwm_pin(unsigned int pin, unsigned int wrap, float clkdiv) {
    gpio_set_function(pin, GPIO_FUNC_PWM);
    unsigned int slice = pwm_gpio_to_slice_num(pin);
    pwm_set_clkdiv(slice, clkdiv);
    pwm_set_wrap(slice, wrap);
    pwm_set_enabled(slice, true);
    pwm_set_gpio_level(pin, 0);
}

void motor_noencoder_init(
    MotorNoEncoder *motor,
    unsigned int lfwd_pin,
    unsigned int rfwd_pin,
    unsigned int lrev_pin,
    unsigned int rrev_pin,
    unsigned int right_fault_pin,
    unsigned int left_fault_pin,
    unsigned int pwm_frequency_hz,
    unsigned int reverse_delay_ms,
    int oc_threshold
) {
    motor->lfwd_pin = lfwd_pin;
    motor->rfwd_pin = rfwd_pin;
    motor->lrev_pin = lrev_pin;
    motor->rrev_pin = rrev_pin;
    motor->right_fault_pin = right_fault_pin;
    motor->left_fault_pin = left_fault_pin;
    motor->reverse_delay_ms = reverse_delay_ms;
    motor->has_last_dir = false;
    motor->overcurrent_right = false;
    motor->overcurrent_left = false;
    motor->oc_count_r = 0;
    motor->oc_count_l = 0;
    motor->oc_threshold = oc_threshold;

    motor->pwm_wrap = 1000000u / pwm_frequency_hz;
    float clkdiv = 125.0f;

    motor_noencoder_setup_pwm_pin(lfwd_pin, motor->pwm_wrap, clkdiv);
    motor_noencoder_setup_pwm_pin(rfwd_pin, motor->pwm_wrap, clkdiv);
    motor_noencoder_setup_pwm_pin(lrev_pin, motor->pwm_wrap, clkdiv);
    motor_noencoder_setup_pwm_pin(rrev_pin, motor->pwm_wrap, clkdiv);

    if (motor_noencoder_fault_pin_valid(right_fault_pin)) {
        gpio_init(right_fault_pin);
        gpio_set_dir(right_fault_pin, GPIO_IN);
        gpio_pull_up(right_fault_pin);
    }
    if (motor_noencoder_fault_pin_valid(left_fault_pin)) {
        gpio_init(left_fault_pin);
        gpio_set_dir(left_fault_pin, GPIO_IN);
        gpio_pull_up(left_fault_pin);
    }
}

void motor_noencoder_check_overcurrent(MotorNoEncoder *motor) {
    if (!motor_noencoder_fault_pin_valid(motor->right_fault_pin)) {
        motor->oc_count_r = 0;
        motor->overcurrent_right = false;
    } else if (gpio_get(motor->right_fault_pin)) {
        motor->oc_count_r++;
        if (motor->oc_count_r >= motor->oc_threshold) {
            motor->overcurrent_right = true;
        }
    } else {
        motor->oc_count_r = 0;
        motor->overcurrent_right = false;
    }

    if (!motor_noencoder_fault_pin_valid(motor->left_fault_pin)) {
        motor->oc_count_l = 0;
        motor->overcurrent_left = false;
    } else if (gpio_get(motor->left_fault_pin)) {
        motor->oc_count_l++;
        if (motor->oc_count_l >= motor->oc_threshold) {
            motor->overcurrent_left = true;
        }
    } else {
        motor->oc_count_l = 0;
        motor->overcurrent_left = false;
    }
}

void motor_noencoder_stop_all(MotorNoEncoder *motor) {
    pwm_set_gpio_level(motor->lfwd_pin, 0);
    pwm_set_gpio_level(motor->rfwd_pin, 0);
    pwm_set_gpio_level(motor->lrev_pin, 0);
    pwm_set_gpio_level(motor->rrev_pin, 0);
}

void motor_noencoder_move(MotorNoEncoder *motor, float speed, bool left_direction, bool right_direction) {
    if (motor->has_last_dir &&
        ((left_direction != motor->last_ldir) || (right_direction != motor->last_rdir))) {
        motor_noencoder_stop_all(motor);
        sleep_ms(motor->reverse_delay_ms);
    }

    if (speed <= 0.0f) {
        motor_noencoder_stop_all(motor);
        return;
    }
    if (speed > 1.0f) {
        speed = 1.0f;
    }

    unsigned int duty = (unsigned int)(speed * motor->pwm_wrap);
    pwm_set_gpio_level(motor->lfwd_pin, 0);
    pwm_set_gpio_level(motor->rfwd_pin, 0);
    pwm_set_gpio_level(motor->lrev_pin, 0);
    pwm_set_gpio_level(motor->rrev_pin, 0);

    if (motor->overcurrent_right) {
        printf("Right Side Overcurrent\n");
    } else if (left_direction) {
        pwm_set_gpio_level(motor->lfwd_pin, duty);
    } else {
        pwm_set_gpio_level(motor->lrev_pin, duty);
    }

    if (motor->overcurrent_left) {
        printf("Left Side Overcurrent\n");
    } else if (right_direction) {
        pwm_set_gpio_level(motor->rfwd_pin, duty);
    } else {
        pwm_set_gpio_level(motor->rrev_pin, duty);
    }

    motor->last_ldir = left_direction;
    motor->last_rdir = right_direction;
    motor->has_last_dir = true;
    printf("xpos: %d"
           "ypos: %d", x_pos, y_pos);
}